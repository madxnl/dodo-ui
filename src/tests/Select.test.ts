import { mount } from '@vue/test-utils'
import { expect, test } from 'vitest'
import { nextTick } from 'vue'
import { Select } from '..'

const _options = [...Array(20)].map((_, i) => ({ label: 'Option ' + i, value: i }))

const getByText = (wrapper: any, text: string) => wrapper.findAll('*').find((e: any) => e.text().trim() === text)!

test('Select a value', async () => {
  const wrapper = mount(Select, {
    props: { options: _options, modelValue: _options[0].value },
    global: { stubs: { teleport: true } },
  })
  await getByText(wrapper, _options[0].label).trigger('click')
  await getByText(wrapper, _options[1].label).trigger('click')
  expect(wrapper.emitted()).toMatchObject({ 'update:modelValue': [[_options[1].value]] })
  expect(wrapper.html()).toMatchSnapshot()
})

test('Select can be disabled', async () => {
  const wrapper = mount(Select, {
    props: { options: _options, modelValue: _options[0].value, disabled: true },
    global: { stubs: { teleport: true } },
  })
  await getByText(wrapper, _options[0].label).trigger('click')
  await nextTick()
  expect(getByText(wrapper, _options[1].label)).toBeFalsy()
  expect(wrapper.html()).toMatchSnapshot()
})

test('Search filters options', async () => {
  const wrapper = mount(Select, {
    props: { options: _options, modelValue: _options[0].value },
    global: { stubs: { teleport: true } },
    attachTo: document.body,
  })
  await getByText(wrapper, _options[0].label).trigger('click')
  await nextTick()
  const search = wrapper.get('input')
  expect(search.element).toBe(document.activeElement)
  search.setValue(_options[2].label)
  await nextTick()
  expect(getByText(wrapper, _options[1].label)).toBeFalsy()
  expect(getByText(wrapper, _options[2].label)).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()
})

test('Search input visible if many options', async () => {
  const wrapper = mount(Select, {
    props: { options: _options.slice(0, 5), modelValue: _options[0].value },
    global: { stubs: { teleport: true } },
  })
  await getByText(wrapper, _options[0].label).trigger('click')
  await nextTick()
  expect(wrapper.find('[type=search]').isVisible()).toBeFalsy()
  wrapper.setProps({ options: _options })
  await nextTick()
  expect(wrapper.get('[type=search]').isVisible()).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()
})

test('Search input visible after character key press', async () => {
  const wrapper = mount(Select, {
    props: { options: _options.slice(0, 5), modelValue: _options[0].value },
    global: { stubs: { teleport: true } },
    attachTo: document.body,
  })
  await getByText(wrapper, _options[0].label).trigger('click')
  await nextTick()
  expect(wrapper.find('[type=search]').isVisible()).toBeFalsy()
  wrapper.get('input').setValue('foo')
  await nextTick()
  expect(wrapper.get('[type=search]').isVisible()).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()
})

// test('Select using keyboard', async () => {
//   const wrapper = mount(Select, {
//     props: { options: _options, modelValue: _options[0].value },
//   })
//   const select = await getByText(wrapper, _options[0].label)
//   await fireEvent.keyDown(select, { key: 'ArrowDown' })
//   await fireEvent.keyDown(select, { key: 'ArrowDown' })
//   await fireEvent.keyDown(select, { key: 'Enter' })
//   expect(emitted()).toMatchObject({ 'update:modelValue': [[_options[1].value]] })
//   expect(wrapper.html()).toMatchSnapshot()
// })

// test('Search by typing and press enter', async () => {
//   const wrapper = mount(Select, {
//     props: { options: _options, modelValue: _options[0].value },
//   })
//   const input = await findByRole('textbox')
//   await getByText(wrapper, _options[0].label)
//   await fireEvent.update(input, _options[1].label)
//   await fireEvent.keyDown(input, { key: 'Enter' })
//   expect(emitted()).toMatchObject({ 'update:modelValue': [[_options[1].value]] })
//   expect(wrapper.html()).toMatchSnapshot()
// })

// test('Custom option html using slot', async () => {
//   const wrapper = mount(Select, {
//     props: { options: _options, modelValue: _options[0].value },
//     slots: { default: option => `Custom ${option.value}` },
//   })
//   await getByText(wrapper, 'Custom 1')
//   await getByText(wrapper, 'Custom 2')
//   expect(emitted()).toMatchObject({ 'update:modelValue': [[_options[1].value]] })
//   expect(wrapper.html()).toMatchSnapshot()
// })

// test('Can clear selection if multiple selected', async () => {
//   const wrapper = mount(Select, {
//     props: { options: _options, modelValue: [_options[0].value], multiple: true },
//   })
//   await getByText(wrapper, _options[0].label)
//   await getByText(wrapper, 'Clear selection')
//   expect(emitted()).toMatchObject({ 'update:modelValue': [[[]]] })
//   expect(wrapper.html()).toMatchSnapshot()
// })
