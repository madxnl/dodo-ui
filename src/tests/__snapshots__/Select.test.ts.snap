// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search filters options 1`] = `
"<div class=\\"_trigger_0c7521\\">
  <div class=\\"_Field_342682 _button_342682\\">
    <div class=\\"_inner_342682\\">
      <p data-nowrap=\\"\\">
        <!--v-if-->Option 0
      </p><input style=\\"position: fixed; pointer-events: none; opacity: 0;\\">
    </div><span class=\\"_icon_db495d material-symbols-outlined\\">arrow_drop_down</span>
  </div>
</div>
<teleport-stub to=\\"body\\">
  <div class=\\"_Dropdown_0c7521\\" style=\\"min-width: 0px; left: 0px; max-width: 76px; top: 1px; max-height: 75px;\\">
    <div class=\\"_el_665c14 dodo-gap-0 dodo-pad-1 dodo-pad-1 _content_0c7521\\">
      <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"min-height: 0;\\">
        <div class=\\"_el_109a5d dodo-pad-2 _searchbox_53a782\\">
          <div class=\\"_Field_342682 _TextInput_6615c1\\" tabindex=\\"0\\" style=\\"flex-grow: 1;\\">
            <div class=\\"_inner_342682\\"><input class=\\"_input_6615c1\\" placeholder=\\"Search\\" type=\\"search\\"></div>
          </div>
        </div>
        <div class=\\"_ScrollContainer_6a6642\\" style=\\"\\">
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 2
              </p>
              <!--v-if-->
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</teleport-stub>"
`;

exports[`Search input visible after character key press 1`] = `
"<div class=\\"_trigger_0c7521\\">
  <div class=\\"_Field_342682 _button_342682\\">
    <div class=\\"_inner_342682\\">
      <p data-nowrap=\\"\\">
        <!--v-if-->Option 0
      </p><input style=\\"position: fixed; pointer-events: none; opacity: 0;\\">
    </div><span class=\\"_icon_db495d material-symbols-outlined\\">arrow_drop_down</span>
  </div>
</div>
<teleport-stub to=\\"body\\">
  <div class=\\"_Dropdown_0c7521\\" style=\\"min-width: 0px; left: 0px; max-width: 76px; top: 1px; max-height: 75px;\\">
    <div class=\\"_el_665c14 dodo-gap-0 dodo-pad-1 dodo-pad-1 _content_0c7521\\">
      <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"min-height: 0;\\">
        <div class=\\"_el_109a5d dodo-pad-2 _searchbox_53a782\\" style=\\"\\">
          <div class=\\"_Field_342682 _TextInput_6615c1\\" tabindex=\\"0\\" style=\\"flex-grow: 1;\\">
            <div class=\\"_inner_342682\\"><input class=\\"_input_6615c1\\" placeholder=\\"Search\\" type=\\"search\\"></div>
          </div>
        </div>
        <div class=\\"_ScrollContainer_6a6642\\" style=\\"\\"></div>
      </div>
    </div>
  </div>
</teleport-stub>"
`;

exports[`Search input visible if many options 1`] = `
"<div class=\\"_trigger_0c7521\\">
  <div class=\\"_Field_342682 _button_342682\\">
    <div class=\\"_inner_342682\\">
      <p data-nowrap=\\"\\">
        <!--v-if-->Option 0
      </p><input style=\\"position: fixed; pointer-events: none; opacity: 0;\\">
    </div><span class=\\"_icon_db495d material-symbols-outlined\\">arrow_drop_down</span>
  </div>
</div>
<teleport-stub to=\\"body\\">
  <div class=\\"_Dropdown_0c7521\\" style=\\"min-width: 0px; left: 0px; max-width: 76px; top: 1px; max-height: 75px;\\">
    <div class=\\"_el_665c14 dodo-gap-0 dodo-pad-1 dodo-pad-1 _content_0c7521\\">
      <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"min-height: 0;\\">
        <div class=\\"_el_109a5d dodo-pad-2 _searchbox_53a782\\">
          <div class=\\"_Field_342682 _TextInput_6615c1\\" tabindex=\\"0\\" style=\\"flex-grow: 1;\\">
            <div class=\\"_inner_342682\\"><input class=\\"_input_6615c1\\" placeholder=\\"Search\\" type=\\"search\\"></div>
          </div>
        </div>
        <div class=\\"_ScrollContainer_6a6642\\" style=\\"\\">
          <!--v-if-->
          <div data-selectactive=\\"true\\" class=\\"_DropdownItem_8d9fa1 _active_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 0
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 1
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 2
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 3
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 4
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 5
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 6
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 7
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 8
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 9
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 10
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 11
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 12
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 13
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 14
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 15
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 16
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 17
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 18
              </p>
              <!--v-if-->
            </div>
          </div>
          <!--v-if-->
          <div class=\\"_DropdownItem_8d9fa1\\">
            <div class=\\"_el_665c14 dodo-gap-0\\" style=\\"flex: 1;\\">
              <p data-nowrap=\\"\\" class=\\"_text_8d9fa1\\">
                <!-- @slot Customize option html -->Option 19
              </p>
              <!--v-if-->
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</teleport-stub>"
`;

exports[`Select a value 1`] = `
"<div class=\\"_trigger_0c7521\\">
  <div class=\\"_Field_342682 _button_342682\\">
    <div class=\\"_inner_342682\\">
      <p data-nowrap=\\"\\">
        <!--v-if-->Option 0
      </p><input style=\\"position: fixed; pointer-events: none; opacity: 0;\\">
    </div><span class=\\"_icon_db495d material-symbols-outlined\\">arrow_drop_down</span>
  </div>
</div>
<teleport-stub to=\\"body\\">
  <!--v-if-->
</teleport-stub>"
`;

exports[`Select can be disabled 1`] = `
"<div class=\\"_trigger_0c7521 _disabled_0c7521\\">
  <div class=\\"_Field_342682 _disabled_342682 _button_342682\\">
    <div class=\\"_inner_342682\\">
      <p data-nowrap=\\"\\">
        <!--v-if-->Option 0
      </p><input style=\\"position: fixed; pointer-events: none; opacity: 0;\\" disabled=\\"\\">
    </div><span class=\\"_icon_db495d material-symbols-outlined\\">arrow_drop_down</span>
  </div>
</div>
<teleport-stub to=\\"body\\">
  <!--v-if-->
</teleport-stub>"
`;
